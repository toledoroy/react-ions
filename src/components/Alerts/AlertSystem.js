import React from 'react'
import PropTypes from 'prop-types'
import style from './style.scss'
import classNames from 'classnames/bind'
import Alert from './Alert'

class AlertSystem extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      alerts: props.alerts
    }
  }

  static propTypes = {
    /**
     * The alerts to display.
     */
    alerts: PropTypes.array.isRequired,
    /**
     * Optional styles to add to the alert system component.
     */
    optClass: PropTypes.string,
    /**
     * Whether or not to slide the alerts in from the right
     */
    slideIn: PropTypes.bool
  }

  getAlerts = () => {
    return this.state.alerts.map((alert, index) =>
      !alert.hidden ? <Alert key={alert.key} type={alert.type || 'success'} optClass={alert.class || ''} closable={typeof alert.closable !== 'undefined' ? alert.closable : true} timeout={alert.timeout} onClose={this.removeAlert.bind(this, alert)}>{alert.content}</Alert> : null
    )
  }

  removeAlert = alert => {
    let alerts = this.state.alerts

    alerts.map((a, index) => {
      if (alert.key === a.key) {
        a.hidden = true

        if (typeof a.onClose === 'function') {
          a.onClose(alert)
        }
      }
    })

    this.setState({ alerts: alerts })
  }

  UNSAFE_componentWillReceiveProps = nextProps => {
    let alerts = nextProps.alerts

    alerts.map((alert, index) => {
      if (!alert.key) {
        alert.key = (alert.type || 'success') + '-' + new Date().getTime()
      }
    })

    this.setState({ alerts: alerts })
  }

  render() {
    const cx = classNames.bind(style)
    const slideInClass = this.props.slideIn ? style['slide-in-right'] : null
    const alertSystemClasses = cx(style['alert-system'], slideInClass, this.props.optClass)

    return (
      <div className={alertSystemClasses}>
        {this.getAlerts()}
      </div>
    )
  }
}

export default AlertSystem
