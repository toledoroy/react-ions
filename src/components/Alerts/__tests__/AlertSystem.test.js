import React from 'react'
import AlertSystem from '../AlertSystem'

describe('AlertSystem', () => {
  describe('render', () => {
    it('should render an empty alert system', () => {
      expect(shallow(<AlertSystem alerts={[]} />)).toMatchSnapshot()
    })

    it('should render an alert system with alerts', () => {
      const clock = sinon.useFakeTimers(1588161287485 * 1000)
      const alerts = [
        { content: 'Test success' },
        { type: 'warning', content: 'Test warning' },
        { type: 'info', content: 'Test info' },
        { type: 'danger', content: 'Test danger' }
      ]
  
      expect(shallow(<AlertSystem alerts={alerts}/>)).toMatchSnapshot()

      clock.restore()
    })

    it('should upadate when the alerts prop changes', () => {
      const clock = sinon.useFakeTimers(1588161287485 * 1000)
      const alerts = [{ content: 'Test success' }]
      const wrapper = shallow(<AlertSystem alerts={alerts}/>)

      expect(wrapper).toMatchSnapshot()

      alerts.push({ type: 'warning', content: 'Test warning' })
      wrapper.setProps({ alerts })

      expect(wrapper).toMatchSnapshot()

      clock.restore()
    })
  })

  describe('getAlerts', () => {
    it('should get alerts that are not hidden', () => {
      const alerts = [
        { content: 'Test success' },
        { type: 'warning', content: 'Test warning', hidden: true },
        { type: 'info', content: 'Test info', hidden: true },
        { type: 'danger', content: 'Test danger' }
      ]
  
      const getAlerts = shallow(<AlertSystem alerts={alerts}/>).instance().getAlerts
      expect(getAlerts()).toMatchSnapshot()
    })
  })

  describe('removeAlert', () => {
    it('should hide an alert and call its onClose callback if provided', () => {
      const alerts = [
        { key: 'ALERT_KEY', content: 'Test success', onClose: sinon.spy() }
      ]
      const wrapper = shallow(<AlertSystem alerts={alerts}/>)
      const removeAlert = wrapper.instance().removeAlert
      const alert = { key: 'ALERT_KEY' }

      expect(wrapper.state('alerts')[0]).toMatchSnapshot()
      
      removeAlert(alert)
      expect(wrapper.state('alerts')[0]).toMatchSnapshot()
      expect(alerts[0].onClose.calledOnce).toBe.true
      expect(alerts[0].onClose.calledWithExactly(alert)).toBe.true
    })
  })
})
