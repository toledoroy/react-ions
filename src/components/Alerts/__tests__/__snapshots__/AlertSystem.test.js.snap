// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertSystem getAlerts should get alerts that are not hidden 1`] = `
Array [
  <Alert
    className=""
    closable={true}
    countdownBar={true}
    onClose={[Function]}
    type="success"
  >
    Test success
  </Alert>,
  false,
  false,
  <Alert
    className=""
    closable={true}
    countdownBar={true}
    onClose={[Function]}
    type="danger"
  >
    Test danger
  </Alert>,
]
`;

exports[`AlertSystem removeAlert should hide an alert and call its onClose callback if provided 1`] = `
Object {
  "content": "Test success",
  "key": "ALERT_KEY",
  "onClose": [Function],
}
`;

exports[`AlertSystem removeAlert should hide an alert and call its onClose callback if provided 2`] = `
Object {
  "content": "Test success",
  "hidden": true,
  "key": "ALERT_KEY",
  "onClose": [Function],
}
`;

exports[`AlertSystem render should render an alert system with alerts 1`] = `
<.sc-bdVaJa
  css={Object {}}
>
  <Alert
    className=""
    closable={true}
    countdownBar={true}
    key="success-1588161287485000"
    onClose={[Function]}
    type="success"
  >
    Test success
  </Alert>
  <Alert
    className=""
    closable={true}
    countdownBar={true}
    key="warning-1588161287485000"
    onClose={[Function]}
    type="warning"
  >
    Test warning
  </Alert>
  <Alert
    className=""
    closable={true}
    countdownBar={true}
    key="info-1588161287485000"
    onClose={[Function]}
    type="info"
  >
    Test info
  </Alert>
  <Alert
    className=""
    closable={true}
    countdownBar={true}
    key="danger-1588161287485000"
    onClose={[Function]}
    type="danger"
  >
    Test danger
  </Alert>
</.sc-bdVaJa>
`;

exports[`AlertSystem render should render an empty alert system 1`] = `
<.sc-bdVaJa
  css={Object {}}
/>
`;

exports[`AlertSystem render should upadate when the alerts prop changes 1`] = `
<.sc-bdVaJa
  css={Object {}}
>
  <Alert
    className=""
    closable={true}
    countdownBar={true}
    key="success-1588161287485000"
    onClose={[Function]}
    type="success"
  >
    Test success
  </Alert>
</.sc-bdVaJa>
`;

exports[`AlertSystem render should upadate when the alerts prop changes 2`] = `
<.sc-bdVaJa
  css={Object {}}
>
  <Alert
    className=""
    closable={true}
    countdownBar={true}
    key="success-1588161287485000"
    onClose={[Function]}
    type="success"
  >
    Test success
  </Alert>
  <Alert
    className=""
    closable={true}
    countdownBar={true}
    key="warning-1588161287485000"
    onClose={[Function]}
    type="warning"
  >
    Test warning
  </Alert>
</.sc-bdVaJa>
`;
